# HG changeset patch
# Parent dbec1f3575481201fc095ad088135b3dcf62c839
convert encoding of environment value when mismatch between Vim and OS

diff --git a/src/os_win32.c b/src/os_win32.c
index dfc0d4d..f19a29f 100644
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -7154,3 +7154,34 @@ fix_arg_enc(void)
     set_alist_count();
 }
 #endif
+
+    int
+mch_setenv(var, value, x)
+    char *var;
+    char *value;
+    int	 x;
+{
+    char_u	*envbuf;
+
+    envbuf = alloc((unsigned)(STRLEN(var) + STRLEN(value) + 2));
+    if (envbuf == NULL)
+	return -1;
+
+    sprintf((char *)envbuf, "%s=%s", var, value);
+
+#ifdef FEAT_MBYTE
+    if (enc_codepage >= 0 && (int)GetACP() != enc_codepage)
+    {
+	WCHAR	    *p = enc_to_utf16(envbuf, NULL);
+
+	vim_free(envbuf);
+	if (p == NULL)
+	    return -1;
+	_wputenv(p);
+    }
+    else
+#endif
+	_putenv((char *)envbuf);
+
+    return 0;
+}
diff --git a/src/os_win32.h b/src/os_win32.h
index a910264..5afcdeb 100644
--- a/src/os_win32.h
+++ b/src/os_win32.h
@@ -203,7 +203,9 @@ Trace(char *pszFormat, ...);
 #define ASSERT_NULL_OR_POINTER(p, type) \
     ASSERT(((p) == NULL)  ||  IsValidAddress((p), sizeof(type), FALSE))
 
-#define mch_setenv(name, val, x) setenv(name, val, x)
+#ifndef HAVE_SETENV
+# define HAVE_SETENV
+#endif
 #define mch_getenv(x) (char_u *)getenv((char *)(x))
 #ifdef __BORLANDC__
 # define vim_mkdir(x, y) mkdir(x)
diff --git a/src/proto/os_win32.pro b/src/proto/os_win32.pro
index 19c59ec..75786bd 100644
--- a/src/proto/os_win32.pro
+++ b/src/proto/os_win32.pro
@@ -63,4 +63,5 @@ void free_cmd_argsW(void);
 void used_file_arg(char *name, int literal, int full_path, int diff_mode);
 void set_alist_count(void);
 void fix_arg_enc(void);
+int mch_setenv(char *var, char *value, int x);
 /* vim: set ft=c : */
